hilti/runtime/src/types/regexp.cc

```diff
@@ -309,256 +311,110 @@ Vector<Bytes> RegExp::findGroups(const Bytes& data) const {
     return groups;
 }
 
-regexp::MatchState RegExp::tokenMatcher() const { return regexp::MatchState(*this); }
-
-// TODO: This is stripped down version of the previous view-based matchig
-// code (see below for original code). Not sure if we still need all of this,
-// or could just call jrx-* functions directly instead of _search_pattern.
-jrx_accept_id RegExp::_search_pattern(jrx_match_state* ms, const Bytes& data, jrx_offset* so, jrx_offset* eo,
-                                      bool find_partial_matches) const {
-    const auto use_stdmatcher = ! _flags.no_sub;
-    const jrx_assertion last = JRX_ASSERTION_EOL | JRX_ASSERTION_EOD;
-    jrx_assertion first = JRX_ASSERTION_BOL | JRX_ASSERTION_BOD;
-
-    jrx_accept_id acc = 0;
-    int8_t need_msdone = 0;
-
-    if ( data.isEmpty() ) {
-        // Nothing to do, but still need to init the match state.
-        jrx_match_state_init(_jrx(), 0, ms);
-        return -1;
-    }
-
-    jrx_offset cur = 0;
-
-    while ( acc <= 0 && cur < data.size() ) {
-        if ( need_msdone ) {
-            jrx_match_state_done(ms);
-
-            if ( jrx_is_anchored(_jrx()) )
-                return 0;
-
-            first = 0;
-        }
-
-        need_msdone = 1;
-
-        jrx_match_state_init(_jrx(), cur, ms);
+std::tuple<int32_t, Bytes> RegExp::find(const Bytes& data) const {
+    assert(_jrx() && "regexp not compiled");
 
-        auto block_start = data.data() + cur;
-        auto block_len = data.size() - cur;
+    auto startp = data.data();
+    auto endp = data.data() + data.size();
 
-#ifdef _DEBUG_MATCHING
-        std::cerr << fmt("feeding |%s| use_stdmatcher=%u first=%u last=%u\n",
-                         escapeBytes(std::string_view((const char*)block_start, block_len)), use_stdmatcher, first,
-                         last);
-#endif
-        auto rc = jrx_regexec_partial(_jrx(), reinterpret_cast<const char*>(block_start), block_len, first, last, ms,
-                                      find_partial_matches);
+    int cur_rc = 0;
+    jrx_offset cur_so = -1;
+    jrx_offset cur_eo = -1;
 
-#ifdef _DEBUG_MATCHING
-        std::cerr << fmt("-> rc=%d ms->offset=%d\n", rc, ms->offset);
-#endif
-        if ( use_stdmatcher && rc == 0 )
-            // No further match.
-            return acc;
+    for ( auto cur = data.data(); cur < endp; cur++ ) {
+        jrx_offset so = -1;
+        jrx_offset eo = -1;
+        jrx_match_state ms;
+        auto rc = _search_pattern(&ms, cur, endp - cur, &so, &eo);
 
         if ( rc > 0 ) {
-            // Match.
-            acc = rc;
+            auto nlen = (eo - so);
+            auto olen = (cur_eo - cur_so);
+            so += (cur - startp);
+            eo += (cur - startp);
```

Comment:
Maybe add a comment (and maybe assertions) here mentioning that even if `so` and `eo` are in uninitialized state this is never off-by-one since `_search_pattern is guaranteed to set them to values >= 0 if there was a match,

```suggestion
           
            // If a match was found `_search_patterns` always initializes `so` and `eo` to valid values.
            assert(so != -1);
            assert(eo != -1);
            so += (cur - startp);
            eo += (cur - startp);
```